api/src/test/java/by/borge/jarl/JarlTests.java

```diff
@@ -1,25 +1,58 @@
 package by.borge.jarl;
 
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
 import java.io.File;
 import java.io.IOException;
+import java.util.List;
 import java.util.Map;
+import java.util.stream.Stream;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
 
 public class JarlTests {
-    @Test
-    void simplePlan() throws IOException {
-        var file = new File("../core/src/test/resources/rego/simple/plan.json");
+    private static Stream<Arguments> testInputAndData() {
+        return Stream.of(
+                Arguments.of(null, null),
+                Arguments.of("foo", null),
+                Arguments.of(null, "bar"),
+                Arguments.of("foo", "bar"),
+                Arguments.of(1337, 42),
+                Arguments.of(13.37, 4.2),
+                Arguments.of(true, false),
+                Arguments.of(false, true),
+                Arguments.of(List.of(), List.of()),
+                Arguments.of(
+                        List.of(1, "two", 3.0),
+                        List.of(4, "five", 6.0)),
+                Arguments.of(
+                        Map.of(
+                                "foo", "bar",
+                                "one", 1,
+                                "list", List.of(1, 2, 3),
+                                "map", Map.of("a", "b")),
+                        Map.of(
+                                "bar", "foo",
+                                "two", 2,
+                                "list", List.of(4, 5, 6),
+                                "map", Map.of("c", "d")))
+        );
+    }
+
+    @ParameterizedTest
```

Comment:
I'm not following here.. I'm taking it this is some JUnit magic :) The method below will be called once for each item in the stream, is that right?

Reply:
Yep. It's all tied together by the test method and the argument factory method having the same name.

I considered using spock/groovy instead, as I find that immensely more readable, but decided against it as this is mainly to test the Java API, and having an extra layer of language indirection in the tests might introduce issues, or even hide issues through the magic of groovy.